pkg/render/monitoring.go

```diff
@@ -0,0 +1,259 @@
+			NodeSelector:           map[string]string{"kubernetes.io/os": "linux"},
+			ServiceAccountName:     "prometheus",
+			ServiceMonitorSelector: &metav1.LabelSelector{MatchLabels: map[string]string{"team": "network-operators"}},
+			PodMonitorSelector:     &metav1.LabelSelector{MatchLabels: map[string]string{"team": "network-operators"}},
+			Version:                components.ComponentPrometheus.Version,
+			Retention:              "24h",
+			Resources:              corev1.ResourceRequirements{Requests: corev1.ResourceList{"memory": resource.MustParse("400Mi")}},
+			RuleSelector: &metav1.LabelSelector{MatchLabels: map[string]string{
+				"prometheus": CalicoNodePrometheus,
+				"role":       "tigera-prometheus-rules",
+			}},
+			Alerting: &monitoringv1.AlertingSpec{
+				Alertmanagers: []monitoringv1.AlertmanagerEndpoints{
+					{
+						Name:      CalicoNodeAlertmanager,
+						Namespace: common.TigeraPrometheusNamespace,
+						Port:      intstr.FromString("web"),
+						Scheme:    "http",
+					},
+				},
+			},
+		},
+	}
+}
+
+func (mc *monitoringComponent) podMonitor() *monitoringv1.PodMonitor {
+	return &monitoringv1.PodMonitor{
+		TypeMeta: metav1.TypeMeta{Kind: monitoringv1.PodMonitorsKind, APIVersion: MonitoringAPIVersion},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      FluentdMetrics,
+			Namespace: common.TigeraPrometheusNamespace,
+			Labels:    map[string]string{"team": "network-operators"},
+		},
+		Spec: monitoringv1.PodMonitorSpec{
+			Selector:          metav1.LabelSelector{MatchLabels: map[string]string{"k8s-app": "fluentd-node"}},
+			NamespaceSelector: monitoringv1.NamespaceSelector{MatchNames: []string{"tigera-fluentd"}},
+			PodMetricsEndpoints: []monitoringv1.PodMetricsEndpoint{
+				{
+					HonorLabels:   true,
+					Interval:      "5s",
+					Port:          "metrics-port",
+					ScrapeTimeout: "5s",
+				},
+			},
+		},
+	}
+}
+
+func (mc *monitoringComponent) prometheusRule() *monitoringv1.PrometheusRule {
+	return &monitoringv1.PrometheusRule{
+		TypeMeta: metav1.TypeMeta{Kind: monitoringv1.PrometheusRuleKind, APIVersion: MonitoringAPIVersion},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      TigeraPrometheusDPRate,
+			Namespace: common.TigeraPrometheusNamespace,
+			Labels: map[string]string{
+				"prometheus": CalicoNodePrometheus,
+				"role":       "tigera-prometheus-rules",
+			},
+		},
+		Spec: monitoringv1.PrometheusRuleSpec{
+			Groups: []monitoringv1.RuleGroup{
+				{
+					Name: "calico.rules",
+					Rules: []monitoringv1.Rule{
+						{
+							Alert:  "DeniedPacketsRate",
+							Expr:   intstr.FromString("rate(calico_denied_packets[10s]) > 50"),
+							Labels: map[string]string{"severity": "critical"},
+							Annotations: map[string]string{
+								"summary":     "Instance {{$labels.instance}} - Large rate of packets denied",
+								"description": "{{$labels.instance}} with calico-node pod {{$labels.pod}} has been denying packets at a fast rate {{$labels.sourceIp}} by policy {{$labels.policy}}.",
```

Comment:
I see that you merely copied this from the custom-resources.yaml. Im wondering what the deal with with what looks like template magic. Should we maybe improve or change these?

Reply:
I think these are Prometheus [templates](https://prometheus.io/docs/prometheus/latest/configuration/template_examples/) in the annotations and labels. I will leave them as-is.